client/hwlib/src/collectors/hardware_info.rs

```diff
@@ -0,0 +1,338 @@
+    Ok(processor)
+}
+
+/// Retrieve CPU information from /proc/cpuinfo
+pub fn collect_processor_info_cpuinfo(
+    cpuinfo_filepath: Option<&'static str>,
+    max_cpu_frequency_filepath: Option<&'static str>,
+) -> Result<devices::Processor, Box<dyn std::error::Error>> {
+    let cpu_info = super::cpuinfo::parse_cpuinfo(cpuinfo_filepath)?;
+    let processor = devices::Processor {
+        codename: String::new(),
+        frequency: super::cpuinfo::read_max_cpu_frequency(max_cpu_frequency_filepath)?,
+        manufacturer: cpu_info.cpu_type,
+        version: cpu_info.model,
+    };
+    Ok(processor)
+}
+
+pub fn collect_chassis_info(
+    smbios_data: &smbioslib::SMBiosData,
+) -> Result<crate::models::devices::Chassis, Box<dyn std::error::Error>> {
+    let chassis_info = &smbios_data.collect::<smbioslib::SMBiosSystemChassisInformation>()[0];
+
+    let chassis_type = chassis_info.chassis_type().ok_or("")?.to_string();
+    let manufacturer = chassis_info.manufacturer().to_string();
+    let sku = chassis_info.sku_number().to_string();
+    let version = chassis_info.version().to_string();
+
+    let chassis = crate::models::devices::Chassis {
+        chassis_type,
+        manufacturer,
+        sku,
+        version,
+    };
+
+    Ok(chassis)
+}
+
+pub fn collect_motherboard_info(
+    smbios_data: &smbioslib::SMBiosData,
+) -> Result<devices::Board, Box<dyn std::error::Error>> {
+    let board_info = &smbios_data.collect::<smbioslib::SMBiosBaseboardInformation>()[0];
+
+    let manufacturer = board_info.manufacturer().to_string();
+    let product_name = board_info.product().to_string();
+    let version = board_info.version().to_string();
+
+    let board = devices::Board {
+        manufacturer,
+        product_name,
+        version,
+    };
+
+    Ok(board)
+}
+
+pub fn collect_motherboard_info_from_device_tree(
+    device_tree_filepath: Option<&'static str>,
+) -> Result<devices::Board, Box<dyn std::error::Error>> {
+    let base_path = match device_tree_filepath {
+        Some(base) => std::path::Path::new(base),
+        None => std::path::Path::new(PROC_DEVICE_TREE_DIRPATH),
+    };
+
+    let manufacturer = std::fs::read_to_string(base_path.join("model"))
+        .map(|s| s.trim().to_string())
+        .unwrap_or_else(|_| "Unknown".to_string());
+
+    let product_name = std::fs::read_to_string(base_path.join("compatible"))
+        .map(|s| s.trim().to_string())
+        .unwrap_or_else(|_| "Unknown".to_string());
+
+    let version = std::fs::read_to_string(base_path.join("model"))
+        .map(|s| s.trim().to_string())
+        .unwrap_or_else(|_| "Unknown".to_string());
+
+    let board = devices::Board {
+        manufacturer,
+        product_name,
+        version,
+    };
+
+    Ok(board)
+}
+
+pub fn get_system_info(
+    smbios_data: &smbioslib::SMBiosData,
+) -> Result<(String, String), Box<dyn std::error::Error>> {
+    let system_info = &smbios_data.collect::<smbioslib::SMBiosSystemInformation>()[0];
```

Comment:
Same comment as above around indexing vs `get`

Reply:
Updated in 9f14753